### PR TITLE
ADD support to models on Grid Gateway

### PR DESCRIPTION
**Pull request by IonesioJunior**
- ADD support to models with grid gateway
- ADD global queries (Could be very useful to know what type of datasets / models we have on our grid network).
- ADD new [notebook](https://github.com/OpenMined/Grid/blob/improve_host_models_support/examples/Serving%20models/4%20-%20Host%20and%20Query%20Model%20with%20Grid.ipynb)

### CODE REVIEW

File: app/websocket/app/main/routes.py [hunk 1]
```diff
@@ -96,6 +96,21 @@ def index():
     return render_template("index.html")
 
 
+@main.route("/dataset-tags", methods=["GET"])
+def get_available_tags():
+    """ Returns all tags stored in this node. Can be very useful to know what datasets this node contains. """
+    available_tags = list()
```
Comment (mari-linhares) [inline app/websocket/app/main/routes.py]: I think this should be a set from the beginning instead of turning it into a set

File: gateway/app/main/routes.py [hunk 2]
```diff
@@ -5,12 +5,14 @@
...
 import json
-
+import random
+import os
 import requests
 
 
 # All grid nodes registered at grid network will be stored here
 grid_nodes = {}
+n_replica = os.getenv("REPLICAS", None)
```
Comment (mari-linhares) [inline gateway/app/main/routes.py]: can we move this to where the other env variables are declared?

Also, it would be nice to have the option of setting this using an argument.

Comment (mari-linhares) [reply gateway/app/main/routes.py]: I think in the future this can actually be supported as an argument to the function, but for now, an env variable/script argument is fine.

File: gateway/app/main/routes.py [hunk 3]
```diff
@@ -52,6 +54,76 @@ def get_connected_nodes():
     )
 
 
+@main.route("/choose-model-host", methods=["GET"])
+def choose_model_host():
+    """ Used to choose some grid node to host a model.
+        PS: Currently we perform this randomly.
+    """
+    if n_replica:
```
Comment (mari-linhares) [inline gateway/app/main/routes.py]: You could just check: 

```
if not n_replica:
   n_replica = 1
```

File: gateway/app/main/routes.py [hunk 4]
```diff
@@ -52,6 +54,76 @@ def get_connected_nodes():
...
+
+    # Invalid body
+    if "model_id" not in body:
+        return Response("", status=400, mimetype="application/json")
+
+    match_nodes = []
+    for node in grid_nodes:
+        try:
+            response = requests.get(grid_nodes[node] + "/models/").content
+        except:
```
Comment (mari-linhares) [inline gateway/app/main/routes.py]: What exception do you expect here? Set this explicitly

File: gateway/app/main/routes.py [hunk 5]
```diff
@@ -52,6 +54,76 @@ def get_connected_nodes():
...
+            match_nodes.append((node, grid_nodes[node]))
+
+    # Return a list[ (id, address) ]  with all grid nodes that have the desired model
+    return Response(json.dumps(match_nodes), status=200, mimetype="application/json")
+
+
+@main.route("/search-available-models", methods=["GET"])
+def available_models():
+    """ Get all available models on the grid network. Can be useful to know what models our grid network have. """
+    models = list()
```
Comment (mari-linhares) [inline gateway/app/main/routes.py]: same comment about sets

File: gateway/app/main/routes.py [hunk 6]
```diff
@@ -52,6 +54,76 @@ def get_connected_nodes():
...
+
+
+@main.route("/search-available-tags", methods=["GET"])
+def available_tags():
+    """ Returns all available tags stored on grid nodes. Can be useful to know what dataset our grid network have. """
+    tags = list()
+    for node in grid_nodes:
+        try:
+            response = requests.get(grid_nodes[node] + "/dataset-tags").content
+        except:
```
Comment (mari-linhares) [inline gateway/app/main/routes.py]: same

File: gateway/app/main/routes.py [hunk 7]
```diff
@@ -52,6 +54,76 @@ def get_connected_nodes():
...
+    models = set(models)
+
+    # Return a list[ "model_id" ]  with all grid nodes
+    return Response(json.dumps(list(models)), status=200, mimetype="application/json")
+
+
+@main.route("/search-available-tags", methods=["GET"])
+def available_tags():
+    """ Returns all available tags stored on grid nodes. Can be useful to know what dataset our grid network have. """
+    tags = list()
```
Comment (mari-linhares) [inline gateway/app/main/routes.py]: same

File: gateway/app/main/routes.py [hunk 8]
```diff
@@ -52,6 +54,76 @@ def get_connected_nodes():
...
+
+
+@main.route("/search-available-models", methods=["GET"])
+def available_models():
+    """ Get all available models on the grid network. Can be useful to know what models our grid network have. """
+    models = list()
+    for node in grid_nodes:
+        try:
+            response = requests.get(grid_nodes[node] + "/models/").content
+        except:
```
Comment (mari-linhares) [inline gateway/app/main/routes.py]: same

File: gateway/app/main/routes.py [hunk 9]
```diff
@@ -64,9 +136,12 @@ def search_dataset_tags():
...
     match_grid_nodes = []
     for node in grid_nodes:
-        response = requests.post(
-            grid_nodes[node] + "/search", data=json.dumps({"query": body["query"]})
-        ).content
+        try:
+            response = requests.post(
+                grid_nodes[node] + "/search", data=json.dumps({"query": body["query"]})
+            ).content
+        except:
```
Comment (mari-linhares) [inline gateway/app/main/routes.py]: same

File: gateway/app/static/styles/index.css [hunk 10]
```diff
@@ -0,0 +1,3 @@
+body {
```
Comment (mari-linhares) [inline gateway/app/static/styles/index.css]: Is this really necessary? Maybe we can make this change in a different PR

File: gateway/app/templates/index.html [hunk 11]
```diff
@@ -1,8 +1,8 @@
 <html>
     <head>
-        <title>Grid Gateway</title>
+	<link rel= "stylesheet" type= "text/css" href= "{{ url_for('static',filename='styles/index.css') }}">
```
Comment (mari-linhares) [inline gateway/app/templates/index.html]: Move this to a different PR

File: grid/grid_network.py [hunk 12]
```diff
@@ -34,16 +34,59 @@ def search(self, *query):
...
-                # There is already a connection to this node
-                worker = sy.hook.local_worker._known_workers[node_id]
-                worker.connect()
+            worker = self.__connect_with_node(node_id, node_url)
             tensor_set.append(worker.search(*query))
         return tensor_set
 
+    def host_model(self, model, model_id):
+        """ This method will choose one of grid nodes registered in the grid network to host
+            host a plain text model.
```
Comment (mari-linhares) [inline grid/grid_network.py]: Fix typo

File: grid/grid_network.py [hunk 13]
```diff
@@ -34,16 +34,59 @@ def search(self, *query):
...
+
+        for host_id, host_address in hosts:
+            # Host model
+            host_worker = self.__connect_with_node(host_id, host_address)
+            host_worker.serve_model(model, model_id=model_id)
+            host_worker.disconnect()
+
+    def query_model(self, model_id):
+        """ This method will search for a specific model registered on grid network, if found,
+            It will return all grid nodes that cointains the desired model.
```
Comment (mari-linhares) [inline grid/grid_network.py]: Fix typo

File: test/test_api_grid.py [hunk 14]
```diff
@@ -39,6 +42,15 @@ def test_connected_nodes(self):
         for node_id in IDS:
             self.assertTrue(node_id in response["grid-nodes"])
 
+    def test_grid_host_query_jit_model(self):
```
Comment (mari-linhares) [inline test/test_api_grid.py]: Can you add the same test for plans?

Comment (IonesioJunior) [reply test/test_api_grid.py]: Done!  (2a4d7aa)

File: grid/grid_network.py (hunk 13 above)
Comment (IonesioJunior) [reply grid/grid_network.py]: Done! (6ace96c)

File: grid/grid_network.py (hunk 12 above)
Comment (IonesioJunior) [reply grid/grid_network.py]: Done! (6ace96c)

File: gateway/app/main/routes.py (hunk 5 above)
Comment (IonesioJunior) [reply gateway/app/main/routes.py]: Done! (6ace96c)

File: gateway/app/main/routes.py (hunk 4 above)
Comment (IonesioJunior) [reply gateway/app/main/routes.py]: Done! (6ace96c)

File: gateway/app/main/routes.py (hunk 3 above)
Comment (IonesioJunior) [reply gateway/app/main/routes.py]: Done! (6ace96c)

File: gateway/app/main/routes.py (hunk 2 above)
Comment (IonesioJunior) [reply gateway/app/main/routes.py]: Done! (6ace96c)

File: app/websocket/app/main/routes.py (hunk 1 above)
Comment (IonesioJunior) [reply app/websocket/app/main/routes.py]: Done! (6ace96c)

File: gateway/app/main/routes.py (hunk 6 above)
Comment (IonesioJunior) [reply gateway/app/main/routes.py]: Done! (6ace96c)

File: gateway/app/main/routes.py (hunk 7 above)
Comment (IonesioJunior) [reply gateway/app/main/routes.py]: Done! (6ace96c)

File: gateway/app/main/routes.py (hunk 8 above)
Comment (IonesioJunior) [reply gateway/app/main/routes.py]: Done! (6ace96c)

File: gateway/app/main/routes.py (hunk 9 above)
Comment (IonesioJunior) [reply gateway/app/main/routes.py]: Done! (6ace96c)

File: gateway/app/templates/index.html (hunk 11 above)
Comment (IonesioJunior) [reply gateway/app/templates/index.html]: Done! (5347e6e)

File: gateway/app/static/styles/index.css (hunk 10 above)
Comment (IonesioJunior) [reply gateway/app/static/styles/index.css]: Done! (5347e6e)

File: gateway/gateway.py [hunk 15]
```diff
@@ -8,7 +8,12 @@
 import os
 
 PORT = os.environ["PORT"]
-app = create_app(debug=False)
+n_replica = os.getenv("REPLICAS", None)
+
+if n_replica:
```
Comment (mari-linhares) [inline gateway/gateway.py]: This if statement is not needed

File: app/websocket/app/main/routes.py [hunk 16]
```diff
@@ -96,6 +96,20 @@ def index():
     return render_template("index.html")
 
 
+@main.route("/dataset-tags", methods=["GET"])
```
Comment (mari-linhares) [inline app/websocket/app/main/routes.py]: can you add a test for this?

File: gateway/gateway.py (hunk 15 above)
Comment (IonesioJunior) [reply gateway/gateway.py]: Done! (ea6da9f)

**Comment by IonesioJunior** [comment]
Added  tests to `/search-available-tags` , `/search-available-models` routes

File: test/test_api_grid.py [hunk 17]
```diff
@@ -39,6 +43,91 @@ def test_connected_nodes(self):
...
+
+            def forward(self, x):
+                x = self.fc1(x)
+                return F.log_softmax(x, dim=0) + self.bias
+
+        model = Net()
+        model.build(th.tensor([1.0, 2]))
+        model_ids = ["first_model", "second_model", "third_model"]
+
+        for i in range(1, len(model_ids) + 1):
```
Comment (mari-linhares) [inline test/test_api_grid.py]: you can just do: `for model_id in model_ids`

File: test/test_api_grid.py [hunk 18]
```diff
@@ -39,6 +43,91 @@ def test_connected_nodes(self):
...
+            self.my_grid.host_model(model, model_id=model_ids[i - 1])
+            models = json.loads(
```
Comment (mari-linhares) [inline test/test_api_grid.py]: you can make a single request after the for

File: test/test_api_grid.py [hunk 19]
```diff
@@ -39,6 +43,91 @@ def test_connected_nodes(self):
...
+                requests.get(GATEWAY_URL + "/search-available-models").content
+            )
+            assert model_ids[i - 1] in model_ids
```
Comment (mari-linhares) [inline test/test_api_grid.py]: and then check if `for model_id in model_ids: assert model_id in models`

Comment (mari-linhares) [inline test/test_api_grid.py]: assert model_ids[i - 1] in model_ids -> assert model_ids[i - 1] in models

File: test/test_api_grid.py [hunk 20]
```diff
@@ -39,6 +43,91 @@ def test_connected_nodes(self):
...
+
+    def test_grid_host_query_jit_model(self):
+        toy_model = th.nn.Linear(2, 5)
+        data = th.zeros((5, 2))
+        traced_model = th.jit.trace(toy_model, data)
+
+        self.my_grid.host_model(traced_model, "test")
+        assert None != self.my_grid.query_model("test")
```
Comment (mari-linhares) [inline test/test_api_grid.py]: assert None != self.my_grid.query_model("test") -> assert self.my_grid.query_model("test")

Comment (mari-linhares) [reply test/test_api_grid.py]: assert None == self.my_grid.query_model("unregistered-model") -> assert self.my_grid.query_model("unregistered-model") is None

File: test/test_api_grid.py [hunk 21]
```diff
@@ -39,6 +43,91 @@ def test_connected_nodes(self):
...
+
+        x = th.tensor([1, 2, 3, 4, 5]).tag("#first-tensor")
+        y = th.tensor([1, 2, 3, 4, 5]).tag("#second-tensor")
+        z = th.tensor([4, 5, 9]).tag(
+            "#generic-tag", "#first-tensor", "#second-tensor"
+        )  # repeat tags
+
+        tensors = [x, y, z]
+        workers = [alice, bob, james]
+        for i in range(1, 4):
```
Comment (mari-linhares) [inline test/test_api_grid.py]: for i in range(len(workers))

File: test/test_api_grid.py [hunk 22]
```diff
@@ -39,6 +43,91 @@ def test_connected_nodes(self):
...
+            tensors[i - 1].send(workers[i - 1])
+            tags = json.loads(
+                requests.get(GATEWAY_URL + "/search-available-tags").content
+            )
+
+            assert i == len(tags)
```
Comment (mari-linhares) [inline test/test_api_grid.py]: assert i + 1 == len(tags)

File: grid/grid_network.py [hunk 23]
```diff
@@ -34,16 +35,72 @@ def search(self, *query):
...
-                worker.connect()
-            else:
-                # There is already a connection to this node
-                worker = sy.hook.local_worker._known_workers[node_id]
-                worker.connect()
+            worker = self.__connect_with_node(node_id, node_url)
             tensor_set.append(worker.search(*query))
         return tensor_set
 
+    def host_model(self, model, model_id):
```
Comment (mari-linhares) [inline grid/grid_network.py]: can you call this serve_model instead? This will imply changing the notebook!

File: test/test_api_grid.py (hunk 19 above)
Comment (IonesioJunior) [reply test/test_api_grid.py]: Isn't it a different suggestion to the same assertion mentioned before?

File: grid/grid_network.py (hunk 23 above)
Comment (IonesioJunior) [reply grid/grid_network.py]: Done! (fa2ea61)

Comment (IonesioJunior) [reply grid/grid_network.py]: Done! (fa9c9bb)

File: test/test_api_grid.py (hunk 22 above)
Comment (IonesioJunior) [reply test/test_api_grid.py]: Done! (9b841da)

File: test/test_api_grid.py (hunk 21 above)
Comment (IonesioJunior) [reply test/test_api_grid.py]: Done! (9b841da)

File: test/test_api_grid.py (hunk 20 above)
Comment (IonesioJunior) [reply test/test_api_grid.py]: Done! (9b841da)

File: test/test_api_grid.py (hunk 19 above)
Comment (IonesioJunior) [reply test/test_api_grid.py]: Done! (9b841da)

File: test/test_api_grid.py (hunk 18 above)
Comment (IonesioJunior) [reply test/test_api_grid.py]: Done! (9b841da)

File: test/test_api_grid.py (hunk 17 above)
Comment (IonesioJunior) [reply test/test_api_grid.py]: Done! (9b841da)
